File: JarvisServer/Services/ClientService.cs
using Microsoft.AspNetCore.SignalR;
using Shared;
using Shared.Messages;

namespace JarvisServer.Services;

public class ClientService(IHubContext<JarvisHub, IUserClient> hub, UserService users, ClientResponseTracker tracker)
{
    public Task SendMessageToAll(string message) =>
        hub.Clients.All.ReceiveMessage(message);

    public async Task<string> SendCommandToUser<T>(AgentMessage<T> message)
        where T : AgentCommand
    {
        try
        {
            Result<string> query = await from id in users.GetConnectionId(message.Key)
                                         let tracking = tracker.Register()
                                         let client = hub.Clients.Client(id)
                                         let task = client.ReceiveCommand(tracking.Id, message.Command)
                                         select task.Select(() => tracking.Task);

            return query.IfError(x => x.Message);
        }
        catch (TaskCanceledException)
        {
            return "Timeout. Client didn't respond.";
        }
    }
}

public static class TaskExtensions
{
    public static async Task<TResult> Select<TResult>(this Task self, Func<Task<TResult>> f)
    {
        await self;
        return await f();
    }
}

File: Server/Services/ClientService.fs
namespace Server.Services

open System.Threading.Tasks
open Common
open Common.SignalR
open Microsoft.AspNetCore.SignalR
open FsToolkit.ErrorHandling
open Microsoft.FSharp.Core

type ClientService(ctx: IHubContext<HubService, IClientService>, users: UserService, tracker: ClientResponseTracker) =
    member this.SendMessageToAll(message) = ctx.Clients.All.ReceiveMessage(message)

    member this.SendCommandToUser(message: AgentMessage) =
        let getId key =
            users.GetConnectionId(key)
            |> ValueOption.toResult $"User not found with key: {message.Key}"

        let core () =
            taskResult {
                let! id = getId message.Key
                let correlationId, trackingTask = tracker.Register()
                let client = ctx.Clients.Client(id)
                do! client.ReceiveCommand(correlationId, message.Command)
                return! trackingTask
            }

        task {
            try
                let! result = core ()
                return result |> Result.defaultWith _.Message
            with :? TaskCanceledException ->
                return "Timeout. Client didn't respond."
        }
